File: clients_sdk/python/bud_foundry/audio/processor.py
"""
Audio processing utilities for Bud Foundry SDK
"""

import struct
from typing import Union


class AudioProcessor:
    """Audio processing utilities for PCM conversion."""

    @staticmethod
    def float32_to_int16(audio: bytes) -> bytes:
        """
        Convert float32 PCM audio to int16 PCM.

        Args:
            audio: Float32 PCM audio bytes

        Returns:
            Int16 PCM audio bytes
        """
        # Unpack as float32
        num_samples = len(audio) // 4
        floats = struct.unpack(f"<{num_samples}f", audio)

        # Convert to int16
        int16_samples = []
        for f in floats:
            # Clamp to [-1, 1]
            f = max(-1.0, min(1.0, f))
            # Scale to int16 range
            int16_samples.append(int(f * 32767))

        return struct.pack(f"<{num_samples}h", *int16_samples)

    @staticmethod
    def int16_to_float32(audio: bytes) -> bytes:
        """
        Convert int16 PCM audio to float32 PCM.

        Args:
            audio: Int16 PCM audio bytes

        Returns:
            Float32 PCM audio bytes
        """
        # Unpack as int16
        num_samples = len(audio) // 2
        int16s = struct.unpack(f"<{num_samples}h", audio)

        # Convert to float32
        floats = [i / 32767.0 for i in int16s]

        return struct.pack(f"<{num_samples}f", *floats)

    @staticmethod
    def resample(
        audio: bytes,
        from_rate: int,
        to_rate: int,
        sample_width: int = 2,
    ) -> bytes:
        """
        Resample audio using linear interpolation.

        Args:
            audio: Audio bytes
            from_rate: Source sample rate
            to_rate: Target sample rate
            sample_width: Bytes per sample (2 for int16)

        Returns:
            Resampled audio bytes
        """
        if from_rate == to_rate:
            return audio

        # Unpack samples
        format_char = "h" if sample_width == 2 else "f"
        num_samples = len(audio) // sample_width
        samples = list(struct.unpack(f"<{num_samples}{format_char}", audio))

        # Calculate resampling ratio
        ratio = to_rate / from_rate
        new_length = int(num_samples * ratio)

        # Linear interpolation
        new_samples = []
        for i in range(new_length):
            pos = i / ratio
            idx = int(pos)
            frac = pos - idx

            if idx + 1 < num_samples:
                sample = samples[idx] * (1 - frac) + samples[idx + 1] * frac
            else:
                sample = samples[idx]

            if sample_width == 2:
                new_samples.append(int(sample))
            else:
                new_samples.append(sample)

        return struct.pack(f"<{new_length}{format_char}", *new_samples)

    @staticmethod
    def stereo_to_mono(audio: bytes, sample_width: int = 2) -> bytes:
        """
        Convert stereo audio to mono by averaging channels.

        Args:
            audio: Stereo audio bytes
            sample_width: Bytes per sample (2 for int16)

        Returns:
            Mono audio bytes
        """
        format_char = "h" if sample_width == 2 else "f"
        num_samples = len(audio) // sample_width
        samples = list(struct.unpack(f"<{num_samples}{format_char}", audio))

        # Average pairs
        mono_samples = []
        for i in range(0, num_samples, 2):
            if i + 1 < num_samples:
                avg = (samples[i] + samples[i + 1]) / 2
            else:
                avg = samples[i]

            if sample_width == 2:
                mono_samples.append(int(avg))
            else:
                mono_samples.append(avg)

        return struct.pack(f"<{len(mono_samples)}{format_char}", *mono_samples)

    @staticmethod
    def chunk_audio(
        audio: bytes,
        chunk_size: int,
        sample_width: int = 2,
    ) -> list[bytes]:
        """
        Split audio into chunks.

        Args:
            audio: Audio bytes
            chunk_size: Size of each chunk in bytes
            sample_width: Bytes per sample

        Returns:
            List of audio chunks
        """
        # Align to sample boundaries
        aligned_size = (chunk_size // sample_width) * sample_width
        if aligned_size == 0:
            aligned_size = sample_width

        chunks = []
        for i in range(0, len(audio), aligned_size):
            chunks.append(audio[i:i + aligned_size])

        return chunks

    @staticmethod
    def calculate_duration(
        audio: bytes,
        sample_rate: int,
        sample_width: int = 2,
        channels: int = 1,
    ) -> float:
        """
        Calculate audio duration in seconds.

        Args:
            audio: Audio bytes
            sample_rate: Sample rate in Hz
            sample_width: Bytes per sample
            channels: Number of channels

        Returns:
            Duration in seconds
        """
        num_samples = len(audio) // (sample_width * channels)
        return num_samples / sample_rate

    @staticmethod
    def calculate_rms(audio: bytes, sample_width: int = 2) -> float:
        """
        Calculate RMS (Root Mean Square) of audio.

        Args:
            audio: Audio bytes
            sample_width: Bytes per sample

        Returns:
            RMS value (0.0 to 1.0 for normalized audio)
        """
        format_char = "h" if sample_width == 2 else "f"
        num_samples = len(audio) // sample_width

        if num_samples == 0:
            return 0.0

        samples = struct.unpack(f"<{num_samples}{format_char}", audio)

        # Calculate RMS
        sum_squares = sum(s * s for s in samples)
        rms: float = float((sum_squares / num_samples) ** 0.5)

        # Normalize to 0-1 range for int16
        if sample_width == 2:
            rms = rms / 32767.0

        return rms

    @staticmethod
    def detect_silence(
        audio: bytes,
        threshold: float = 0.01,
        sample_width: int = 2,
    ) -> bool:
        """
        Detect if audio chunk is silence.

        Args:
            audio: Audio bytes
            threshold: RMS threshold below which is considered silence
            sample_width: Bytes per sample

        Returns:
            True if audio is silence
        """
        rms = AudioProcessor.calculate_rms(audio, sample_width)
        return rms < threshold

    @staticmethod
    def normalize(
        audio: bytes,
        target_level: float = 0.9,
        sample_width: int = 2,
    ) -> bytes:
        """
        Normalize audio to a target level.

        Args:
            audio: Audio bytes
            target_level: Target peak level (0.0 to 1.0)
            sample_width: Bytes per sample

        Returns:
            Normalized audio bytes
        """
        format_char = "h" if sample_width == 2 else "f"
        num_samples = len(audio) // sample_width

        if num_samples == 0:
            return audio

        samples = list(struct.unpack(f"<{num_samples}{format_char}", audio))

        # Find peak
        max_val = 32767.0 if sample_width == 2 else 1.0
        peak = max(abs(s) for s in samples) if samples else 0

        if peak == 0:
            return audio

        # Calculate gain
        target_peak = target_level * max_val
        gain = target_peak / peak

        # Apply gain
        normalized = []
        for s in samples:
            new_val = s * gain
            if sample_width == 2:
                new_val = max(-32768, min(32767, int(new_val)))
            normalized.append(new_val)

        return struct.pack(f"<{num_samples}{format_char}", *normalized)

File: clients_sdk/typescript/src/audio/processor.ts
/**
 * Audio Processor
 * Handles audio format conversions for the SDK
 */

/**
 * Audio format information
 */
export interface AudioFormat {
  /** Sample rate in Hz */
  sampleRate: number;
  /** Number of channels */
  channels: number;
  /** Bits per sample (16 for Int16, 32 for Float32) */
  bitsPerSample: 16 | 32;
  /** Whether data is floating point */
  isFloat: boolean;
}

/**
 * Default audio formats
 */
export const AUDIO_FORMATS = {
  /** Standard STT input format */
  STT_INPUT: { sampleRate: 16000, channels: 1, bitsPerSample: 16, isFloat: false } as AudioFormat,
  /** Standard TTS output format */
  TTS_OUTPUT: { sampleRate: 24000, channels: 1, bitsPerSample: 16, isFloat: false } as AudioFormat,
  /** Web Audio API format */
  WEB_AUDIO: { sampleRate: 48000, channels: 1, bitsPerSample: 32, isFloat: true } as AudioFormat,
} as const;

/**
 * Audio processor for format conversions
 */
export class AudioProcessor {
  /**
   * Convert Float32Array to Int16Array (for sending to server)
   * @param float32 Input audio in Float32 format (range -1.0 to 1.0)
   * @returns Audio in Int16 format (range -32768 to 32767)
   */
  static float32ToInt16(float32: Float32Array): Int16Array {
    const int16 = new Int16Array(float32.length);
    for (let i = 0; i < float32.length; i++) {
      // Clamp to -1.0 to 1.0 range
      const sample = Math.max(-1, Math.min(1, float32[i]!));
      // Convert to 16-bit integer
      int16[i] = sample < 0 ? sample * 32768 : sample * 32767;
    }
    return int16;
  }

  /**
   * Convert Int16Array to Float32Array (for Web Audio API)
   * @param int16 Input audio in Int16 format (range -32768 to 32767)
   * @returns Audio in Float32 format (range -1.0 to 1.0)
   */
  static int16ToFloat32(int16: Int16Array): Float32Array {
    const float32 = new Float32Array(int16.length);
    for (let i = 0; i < int16.length; i++) {
      const sample = int16[i]!;
      // Convert from 16-bit integer to float
      float32[i] = sample < 0 ? sample / 32768 : sample / 32767;
    }
    return float32;
  }

  /**
   * Convert ArrayBuffer containing Int16 PCM to Float32Array
   * @param buffer ArrayBuffer containing Int16 PCM data
   * @returns Float32Array for Web Audio API
   */
  static pcmBufferToFloat32(buffer: ArrayBuffer): Float32Array {
    const int16 = new Int16Array(buffer);
    return AudioProcessor.int16ToFloat32(int16);
  }

  /**
   * Convert Float32Array to ArrayBuffer containing Int16 PCM
   * @param float32 Float32Array audio data
   * @returns ArrayBuffer with Int16 PCM data
   */
  static float32ToPcmBuffer(float32: Float32Array): ArrayBuffer {
    const int16 = AudioProcessor.float32ToInt16(float32);
    return int16.buffer as ArrayBuffer;
  }

  /**
   * Resample audio to a different sample rate
   * Uses linear interpolation for simplicity
   * @param input Input audio data
   * @param inputSampleRate Input sample rate in Hz
   * @param outputSampleRate Output sample rate in Hz
   * @returns Resampled audio data
   */
  static resample(input: Float32Array, inputSampleRate: number, outputSampleRate: number): Float32Array {
    if (inputSampleRate === outputSampleRate) {
      return input;
    }

    const ratio = inputSampleRate / outputSampleRate;
    const outputLength = Math.ceil(input.length / ratio);
    const output = new Float32Array(outputLength);

    for (let i = 0; i < outputLength; i++) {
      const srcIndex = i * ratio;
      const srcIndexInt = Math.floor(srcIndex);
      const srcIndexFrac = srcIndex - srcIndexInt;

      if (srcIndexInt + 1 < input.length) {
        // Linear interpolation between two samples
        output[i] = input[srcIndexInt]! * (1 - srcIndexFrac) + input[srcIndexInt + 1]! * srcIndexFrac;
      } else if (srcIndexInt < input.length) {
        output[i] = input[srcIndexInt]!;
      } else {
        output[i] = 0;
      }
    }

    return output;
  }

  /**
   * Resample Int16 audio
   */
  static resampleInt16(input: Int16Array, inputSampleRate: number, outputSampleRate: number): Int16Array {
    const float32 = AudioProcessor.int16ToFloat32(input);
    const resampled = AudioProcessor.resample(float32, inputSampleRate, outputSampleRate);
    return AudioProcessor.float32ToInt16(resampled);
  }

  /**
   * Merge multiple audio channels into mono
   * @param channels Array of channel data (each Float32Array)
   * @returns Mono audio (average of all channels)
   */
  static mergeToMono(channels: Float32Array[]): Float32Array {
    if (channels.length === 0) {
      return new Float32Array(0);
    }

    if (channels.length === 1) {
      return channels[0]!;
    }

    const length = Math.min(...channels.map((c) => c.length));
    const mono = new Float32Array(length);

    for (let i = 0; i < length; i++) {
      let sum = 0;
      for (const channel of channels) {
        sum += channel[i]!;
      }
      mono[i] = sum / channels.length;
    }

    return mono;
  }

  /**
   * Split stereo audio into separate channels
   * @param interleaved Interleaved stereo audio (L R L R L R ...)
   * @returns Array of two Float32Arrays [left, right]
   */
  static splitStereo(interleaved: Float32Array): [Float32Array, Float32Array] {
    const length = Math.floor(interleaved.length / 2);
    const left = new Float32Array(length);
    const right = new Float32Array(length);

    for (let i = 0; i < length; i++) {
      left[i] = interleaved[i * 2]!;
      right[i] = interleaved[i * 2 + 1]!;
    }

    return [left, right];
  }

  /**
   * Calculate RMS (Root Mean Square) level of audio
   * @param audio Audio data
   * @returns RMS level (0.0 to 1.0 for normalized audio)
   */
  static calculateRMS(audio: Float32Array): number {
    if (audio.length === 0) return 0;

    let sumSquares = 0;
    for (let i = 0; i < audio.length; i++) {
      sumSquares += audio[i]! * audio[i]!;
    }

    return Math.sqrt(sumSquares / audio.length);
  }

  /**
   * Calculate peak level of audio
   * @param audio Audio data
   * @returns Peak level (0.0 to 1.0 for normalized audio)
   */
  static calculatePeak(audio: Float32Array): number {
    if (audio.length === 0) return 0;

    let peak = 0;
    for (let i = 0; i < audio.length; i++) {
      const abs = Math.abs(audio[i]!);
      if (abs > peak) peak = abs;
    }

    return peak;
  }

  /**
   * Convert decibels to linear amplitude
   * @param db Decibel value
   * @returns Linear amplitude
   */
  static dbToLinear(db: number): number {
    return Math.pow(10, db / 20);
  }

  /**
   * Convert linear amplitude to decibels
   * @param linear Linear amplitude
   * @returns Decibel value
   */
  static linearToDb(linear: number): number {
    if (linear <= 0) return -Infinity;
    return 20 * Math.log10(linear);
  }

  /**
   * Apply gain to audio
   * @param audio Audio data
   * @param gainDb Gain in decibels
   * @returns Audio with gain applied
   */
  static applyGain(audio: Float32Array, gainDb: number): Float32Array {
    const gainLinear = AudioProcessor.dbToLinear(gainDb);
    const output = new Float32Array(audio.length);

    for (let i = 0; i < audio.length; i++) {
      output[i] = audio[i]! * gainLinear;
    }

    return output;
  }

  /**
   * Normalize audio to target peak level
   * @param audio Audio data
   * @param targetPeakDb Target peak level in dB (default -1 dB)
   * @returns Normalized audio
   */
  static normalize(audio: Float32Array, targetPeakDb = -1): Float32Array {
    const currentPeak = AudioProcessor.calculatePeak(audio);
    if (currentPeak === 0) return audio;

    const targetPeakLinear = AudioProcessor.dbToLinear(targetPeakDb);
    const gain = targetPeakLinear / currentPeak;

    const output = new Float32Array(audio.length);
    for (let i = 0; i < audio.length; i++) {
      output[i] = audio[i]! * gain;
    }

    return output;
  }

  /**
   * Concatenate multiple audio buffers
   * @param buffers Array of Float32Arrays to concatenate
   * @returns Single concatenated Float32Array
   */
  static concatenate(buffers: Float32Array[]): Float32Array {
    const totalLength = buffers.reduce((sum, buf) => sum + buf.length, 0);
    const result = new Float32Array(totalLength);

    let offset = 0;
    for (const buffer of buffers) {
      result.set(buffer, offset);
      offset += buffer.length;
    }

    return result;
  }

  /**
   * Slice audio buffer
   * @param audio Audio data
   * @param startSample Start sample index
   * @param endSample End sample index (exclusive)
   * @returns Sliced audio
   */
  static slice(audio: Float32Array, startSample: number, endSample?: number): Float32Array {
    return audio.slice(startSample, endSample);
  }

  /**
   * Create silence of specified duration
   * @param sampleRate Sample rate in Hz
   * @param durationMs Duration in milliseconds
   * @returns Float32Array of silence
   */
  static createSilence(sampleRate: number, durationMs: number): Float32Array {
    const samples = Math.ceil((sampleRate * durationMs) / 1000);
    return new Float32Array(samples);
  }

  /**
   * Detect if audio contains silence (based on RMS threshold)
   * @param audio Audio data
   * @param thresholdDb Threshold in dB (default -40 dB)
   * @returns True if audio is below threshold
   */
  static isSilence(audio: Float32Array, thresholdDb = -40): boolean {
    const rms = AudioProcessor.calculateRMS(audio);
    const rmsDb = AudioProcessor.linearToDb(rms);
    return rmsDb < thresholdDb;
  }
}
